### ISSUE TITLE
Silent Failure With UnitTest Reference

### ISSUE THREAD

**Issue by nnpcYvIVl** [SonarSource/sonar-dotnet#2353]
Many analyzers silently fail solution-wide when I add a reference to ```Microsoft.VisualStudio.QualityTools.UnitTestFramework.dll```.

In the following sample code S1481 is flagged but S1643 isn't.

```chsarp
namespace SonarLint_Silent_Failure_With_UnitTest_Reference
{
    class Class
    {
        void S1481()
        {
            int x;
        }

        void S1643()
        {
            string s = string.Empty;

            for (int i = 0; i < 10; i++)
            {
                s += i.ToString();
            }
        }
    }
}
```

Here are a list of others that are failing:
```
S1075
S1144
S1168
S1186
S1643
S1871
S2201
S2259
S2327
S2365
S3168
S3869
S3925
S4200
S4214
```

### Related information

* SonarC# Version: Latest
* Visual Studio Version: 15.9.9

**Comment by Evangelink** [comment]
Hi @nnpcYvIVl,

Our analyzer detects whether it is running against some "code" or some "test code" and by default turns on/off some rules. This is why you have this behavior.

**Comment by duncanp-sonar** [comment]
@nnpcYvIVl the SonarC# rules currently look at the project as a whole to decide whether it is a test project or not. A project is regarded as a test project if it references a [fixed list](https://github.com/SonarSource/sonar-dotnet/blob/master/sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/ProjectTypeHelper.cs#L56) of test framework assemblies.

If a project is regarded as a test project then only rules marked as applying to "Test sources" will be executed. [S1481](https://jira.sonarsource.com/browse/RSPEC-1481) is marked as applying to "Main Sources" and "Test Sources" (see the _Analysis Scope_ field in the _Field Tab_) so it is being run. [S1643](https://jira.sonarsource.com/browse/RSPEC-1643) is only marked as applying to "Main Sources" so it isn't executed.

That's why you are getting the behaviour you are seeing. We understand that this isn't helpful in your scenario. Unfortunately, there isn't currently any way to work round this behaviour.

We do have items in the backlog to improve how we handle the categorisation and analysis of test code (as this also impacts the Scanner for MSBuild and SonarLint for Visual Studio), but we don't currently have any firm plans about when this will be tackled.

Out of interest, why do you have lots of projects referencing the test framework? Are the projects part of your test infrastructure, or is your product related to unit testing?

**Comment by nnpcYvIVl** [comment]
@duncanp-sonar Thanks for the detailed explanation.

Some of our developers use an "all-in-one" project style where main and test code are stored together and separated with preprocessor directives. As you described the analyzer detects the presence of a unit test dll and disables a subset of the analyzers.

For the projects in question and the sample code above it's not clear from the developer's perspective that a subset of the analyzers have been disabled. This ends up creating confusion and causes developers to question whether the analyzer is working or not.

**Comment by andrei-epure-sonarsource** [comment]
> Some of our developers use an "all-in-one" project style where main and test code are stored together and separated with preprocessor directives.

I think this is quite an edge case.

Regarding this, maybe we can improve the user experience by adding more logging to raise the awareness...

**Comment by pavel-mikula-sonarsource** [comment]
I agree with "More logging" approach as part of "Test as 1st class citizen" objective.

**Comment by Bouke** [comment]
When a project is categorised as a test project, it apparently also disables any roslyn analyzers that were explicitly added to such projects. I think this is counterintuitive and not at all desirable. Is there a way to opt out of this behaviour?

**Comment by duncanp-sonar** [comment]
@Bouke this is currently no way to opt out of this behaviour. Your only option would be to [explicitly mark](https://github.com/SonarSource/sonar-scanner-msbuild/wiki/Analysis-of-product-projects-vs.-test-projects#explicit-setting-the-project-type) the projects as non-test projects.

**Comment by pavel-mikula-sonarsource** [comment]
I don't see an action that we can take now. 
S4NET run does quite some logging about project type detection already.
There's no place to log output when using analyzers as NuGet packages.
